portfolio/src/main/java/com/google/sps/data/HighestPriorityFirstScheduler.java

```diff
@@ -0,0 +1,118 @@
+package com.google.sps.data;
+
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+
+public class HighestPriorityFirstScheduler implements TaskScheduler {
+  private static final Comparator<ScheduledTask> sortByScheduledStartTimeAscending =
+      Comparator.comparing(ScheduledTask::getStartTime);
+
+  /** This method schedules tasks from highest to lowest priority and */
+  public Collection<ScheduledTask> schedule(
+      Collection<CalendarEvent> events,
+      Collection<Task> tasks,
+      Instant workHoursStartTime,
+      Instant workHoursEndTime) {
+
+    List<CalendarEvent> eventsList = new ArrayList<CalendarEvent>(events);
+    List<Task> tasksList = new ArrayList<Task>(tasks);
+
+    TaskQueue taskQueue = new TaskQueue(tasksList, getSchedulingAlgorithmType());
+
+    CalendarEventsGroup calendarEventsGroup =
+        new CalendarEventsGroup(eventsList, workHoursStartTime, workHoursEndTime);
+
+    List<TimeRange> availableTimes = calendarEventsGroup.getFreeTimeRanges();
+    TimeRangeGroup availableTimesGroup = new ArrayListTimeRangeGroup(availableTimes);
+    List<TimeRange> currentAvailableTimes = getAvailableTimeRangesList(availableTimesGroup);
```

Comment:
The reasoning is that I want to be able to index into currentAvailableTimes in the algorithm since I won't want to get the next availableTime from availableTimesGroup during each loop, so in some cases the index is increased and in some cases it remains the same. I had previously implemented the algorithm using the availableTimesGroup iterator but I had to use a bool to determine whether or not I would call next() on the iterator and it seemed a bit hacky to me.

Reply:
I'll take any suggestions into using the iterator.

Reply:
I think you don't have to use the iterator. But won't availableTimes be the same as currentAvailableTimes here? So maybe we can just assign currentAvailableTimes to availableTimes.

Reply:
Done.